// @sphereon/did-uni-client/dist/resolver/UniResolver.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getUniResolvers = exports.getUniResolver = exports.UniResolver = void 0;
const cross_fetch_1 = __importDefault(require("cross-fetch"));
const did_resolver_1 = require("did-resolver");
const constants_1 = require("../types/constants");
/**
 * Class for performing various DID resolving operations.
 */
class UniResolver {
    /** Resolver constructor */
    constructor(config) {
        this.config = Object.assign(Object.assign({}, constants_1.DefaultConfig), config);
    }
    /**
     * Sets the base URL for the registrar.
     *
     * @param url The base URL for the registrar.
     * @return this.
     */
    setBaseURL(url) {
        const path = constants_1.Constants.URL_PATHNAME_REGEX.exec(this.config.resolveURL);
        if (path) {
            this.config.resolveURL = `${url}${path[1]}`;
        }
        return this;
    }
    /**
     * Sets the URL for the resolve endpoint.
     *
     * @param url The URL for the resolve endpoint.
     * @return this.
     */
    setResolveURL(url) {
        this.config.resolveURL = url;
        return this;
    }
    /**
     * Gets the configuration.
     * @return Config The config.
     */
    getConfig() {
        return this.config;
    }
    /**
     * Resolves a given did to did document.
     *
     * @param did The identifier (did).
     * @return Promise<DIDResolutionResult>, resolution result.
     */
    resolve(didUrl, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const parsedDid = (0, did_resolver_1.parse)(didUrl);
            if (parsedDid === null) {
                return (0, constants_1.errorResolutionResult)('invalidDid');
            }
            const url = `${(options === null || options === void 0 ? void 0 : options.resolveUrl) ? options.resolveUrl : this.config.resolveURL}/${parsedDid.did}`;
            const response = yield (0, cross_fetch_1.default)(url);
            if (response.status >= 400) {
                throw new Error(yield response.text());
            }
            else {
                return yield response.json();
            }
        });
    }
}
exports.UniResolver = UniResolver;
/**
 * Packaging the resolver as a driver to meet https://github.com/decentralized-identity/did-resolver spec
 */
function getUniResolver(didMethod, opts) {
    if (!didMethod) {
        throw new Error('Please provide a did method for the uni-resolver client to resolve a DID document for using the method');
    }
    const uniResolver = new UniResolver();
    if (opts === null || opts === void 0 ? void 0 : opts.resolveUrl) {
        uniResolver.setResolveURL(opts.resolveUrl);
    }
    if (opts === null || opts === void 0 ? void 0 : opts.baseUrl) {
        uniResolver.setBaseURL(opts.baseUrl);
    }
    /* eslint-disable @typescript-eslint/no-unused-vars */
    function resolve(did, _parsed, _resolver, _options) {
        return __awaiter(this, void 0, void 0, function* () {
            return uniResolver.resolve(did, opts);
        });
    }
    /* eslint-enable @typescript-eslint/no-unused-vars */
    return { [didMethod]: resolve };
}
exports.getUniResolver = getUniResolver;
function getUniResolvers(didMethods, opts) {
    if (!didMethods || didMethods.length == 0) {
        throw new Error('Please provide at least one DID method for the uni-resolver client to resolve a DID document for using the method');
    }
    return didMethods.map((didMethod) => getUniResolver(didMethod, opts));
}
exports.getUniResolvers = getUniResolvers;
