// @sphereon/oid4vci-common/dist/functions/RandomUtils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.assertValidCodeVerifier = exports.createCodeChallenge = exports.generateCodeVerifier = exports.generateNonce = exports.generateRandomString = exports.NONCE_LENGTH = exports.CODE_VERIFIER_DEFAULT_LENGTH = void 0;
const oid4vc_common_1 = require("@sphereon/oid4vc-common");
const u8a = __importStar(require("uint8arrays"));
const types_1 = require("../types");
const randomBytes_1 = require("./randomBytes");
exports.CODE_VERIFIER_DEFAULT_LENGTH = 128;
exports.NONCE_LENGTH = 32;
const generateRandomString = (length, encoding) => {
    return u8a.toString((0, randomBytes_1.randomBytes)(length), encoding).slice(0, length);
};
exports.generateRandomString = generateRandomString;
const generateNonce = (length) => {
    return (0, exports.generateRandomString)(length !== null && length !== void 0 ? length : exports.NONCE_LENGTH);
};
exports.generateNonce = generateNonce;
const generateCodeVerifier = (length) => {
    const codeVerifier = (0, exports.generateRandomString)(length !== null && length !== void 0 ? length : exports.CODE_VERIFIER_DEFAULT_LENGTH, 'base64url');
    (0, exports.assertValidCodeVerifier)(codeVerifier);
    return codeVerifier;
};
exports.generateCodeVerifier = generateCodeVerifier;
const createCodeChallenge = (codeVerifier, codeChallengeMethod) => {
    if (codeChallengeMethod === types_1.CodeChallengeMethod.plain) {
        return codeVerifier;
    }
    else if (!codeChallengeMethod || codeChallengeMethod === types_1.CodeChallengeMethod.S256) {
        return u8a.toString((0, oid4vc_common_1.defaultHasher)(codeVerifier, 'sha256'), 'base64url');
    }
    else {
        // Just a precaution if a new method would be introduced
        throw Error(`code challenge method ${codeChallengeMethod} not implemented`);
    }
};
exports.createCodeChallenge = createCodeChallenge;
const assertValidCodeVerifier = (codeVerifier) => {
    const length = codeVerifier.length;
    if (length < 43) {
        throw Error(`code_verifier should have a minimum length of 43; see rfc7636`);
    }
    else if (length > 128) {
        throw Error(`code_verifier should have a maximum length of 128; see rfc7636`);
    }
};
exports.assertValidCodeVerifier = assertValidCodeVerifier;
