// @credo-ts/openid4vc/build/openid4vc-issuer/OpenId4VcIssuerModuleConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenId4VcIssuerModuleConfig = void 0;
const router_1 = require("../shared/router");
const DEFAULT_C_NONCE_EXPIRES_IN = 5 * 60; // 5 minutes
const DEFAULT_TOKEN_EXPIRES_IN = 3 * 60; // 3 minutes
const DEFAULT_PRE_AUTH_CODE_EXPIRES_IN = 3 * 60; // 3 minutes
class OpenId4VcIssuerModuleConfig {
    constructor(options) {
        this.options = options;
        this.router = options.router ?? (0, router_1.importExpress)().Router();
    }
    get baseUrl() {
        return this.options.baseUrl;
    }
    /**
     * Get the credential endpoint config, with default values set
     */
    get credentialEndpoint() {
        // Use user supplied options, or return defaults.
        const userOptions = this.options.endpoints.credential;
        return {
            ...userOptions,
            endpointPath: userOptions.endpointPath ?? '/credential',
        };
    }
    /**
     * Get the access token endpoint config, with default values set
     */
    get accessTokenEndpoint() {
        // Use user supplied options, or return defaults.
        const userOptions = this.options.endpoints.accessToken ?? {};
        return {
            ...userOptions,
            endpointPath: userOptions.endpointPath ?? '/token',
            cNonceExpiresInSeconds: userOptions.cNonceExpiresInSeconds ?? DEFAULT_C_NONCE_EXPIRES_IN,
            preAuthorizedCodeExpirationInSeconds: userOptions.preAuthorizedCodeExpirationInSeconds ?? DEFAULT_PRE_AUTH_CODE_EXPIRES_IN,
            tokenExpiresInSeconds: userOptions.tokenExpiresInSeconds ?? DEFAULT_TOKEN_EXPIRES_IN,
        };
    }
    /**
     * Get the hosted credential offer endpoint config, with default values set
     */
    get credentialOfferEndpoint() {
        // Use user supplied options, or return defaults.
        const userOptions = this.options.endpoints.credentialOffer ?? {};
        return {
            ...userOptions,
            endpointPath: userOptions.endpointPath ?? '/offers',
        };
    }
}
exports.OpenId4VcIssuerModuleConfig = OpenId4VcIssuerModuleConfig;
