// @sphereon/ssi-sdk-ext.identifier-resolution/dist/types/externalIdentifierTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isExternalIdentifierX5cOpts = exports.isExternalIdentifierJwksUrlOpts = exports.isExternalIdentifierOidcDiscoveryOpts = exports.isExternalIdentifierCoseKeyOpts = exports.isExternalIdentifierJwkOpts = exports.isExternalIdentifierKidOpts = exports.isExternalIdentifierDidOpts = void 0;
const common_1 = require("./common");
function isExternalIdentifierDidOpts(opts) {
    const { identifier } = opts;
    return ('method' in opts && opts.method === 'did') || (0, common_1.isDidIdentifier)(identifier);
}
exports.isExternalIdentifierDidOpts = isExternalIdentifierDidOpts;
function isExternalIdentifierKidOpts(opts) {
    const { identifier } = opts;
    return ('method' in opts && opts.method === 'kid') || (0, common_1.isKidIdentifier)(identifier);
}
exports.isExternalIdentifierKidOpts = isExternalIdentifierKidOpts;
function isExternalIdentifierJwkOpts(opts) {
    const { identifier } = opts;
    return ('method' in opts && opts.method === 'jwk') || (0, common_1.isJwkIdentifier)(identifier);
}
exports.isExternalIdentifierJwkOpts = isExternalIdentifierJwkOpts;
function isExternalIdentifierCoseKeyOpts(opts) {
    const { identifier } = opts;
    return ('method' in opts && opts.method === 'cose_key') || (0, common_1.isCoseKeyIdentifier)(identifier);
}
exports.isExternalIdentifierCoseKeyOpts = isExternalIdentifierCoseKeyOpts;
function isExternalIdentifierOidcDiscoveryOpts(opts) {
    const { identifier } = opts;
    return ('method' in opts && opts.method === 'oidc-discovery') || (0, common_1.isOidcDiscoveryIdentifier)(identifier);
}
exports.isExternalIdentifierOidcDiscoveryOpts = isExternalIdentifierOidcDiscoveryOpts;
function isExternalIdentifierJwksUrlOpts(opts) {
    const { identifier } = opts;
    return ('method' in opts && opts.method === 'oidc-discovery') || (0, common_1.isJwksUrlIdentifier)(identifier);
}
exports.isExternalIdentifierJwksUrlOpts = isExternalIdentifierJwksUrlOpts;
function isExternalIdentifierX5cOpts(opts) {
    const { identifier } = opts;
    return ('method' in opts && opts.method === 'x5c') || (0, common_1.isX5cIdentifier)(identifier);
}
exports.isExternalIdentifierX5cOpts = isExternalIdentifierX5cOpts;
