// @sphereon/ssi-sdk-ext.identifier-resolution/dist/types/managedIdentifierTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isManagedIdentifierKeyResult = exports.isManagedIdentifierKidResult = exports.isManagedIdentifierJwkResult = exports.isManagedIdentifierX5cResult = exports.isManagedIdentifierDidResult = exports.isManagedIdentifierCoseKeyResult = exports.isManagedIdentifierX5cOpts = exports.isManagedIdentifierJwkOpts = exports.isManagedIdentifierOID4VCIssuerOpts = exports.isManagedIdentifierCoseKeyOpts = exports.isManagedIdentifierKeyOpts = exports.isManagedIdentifierKidOpts = exports.isManagedIdentifierDidOpts = void 0;
const common_1 = require("./common");
function isManagedIdentifierDidOpts(opts) {
    const { identifier } = opts;
    return ('method' in opts && opts.method === 'did') || (0, common_1.isDidIdentifier)(identifier);
}
exports.isManagedIdentifierDidOpts = isManagedIdentifierDidOpts;
function isManagedIdentifierKidOpts(opts) {
    const { identifier } = opts;
    return ('method' in opts && opts.method === 'kid') || (0, common_1.isKidIdentifier)(identifier);
}
exports.isManagedIdentifierKidOpts = isManagedIdentifierKidOpts;
function isManagedIdentifierKeyOpts(opts) {
    const { identifier } = opts;
    return ('method' in opts && opts.method === 'key') || (0, common_1.isKeyIdentifier)(identifier);
}
exports.isManagedIdentifierKeyOpts = isManagedIdentifierKeyOpts;
function isManagedIdentifierCoseKeyOpts(opts) {
    const { identifier } = opts;
    return ('method' in opts && opts.method === 'cose_key') || (0, common_1.isCoseKeyIdentifier)(identifier);
}
exports.isManagedIdentifierCoseKeyOpts = isManagedIdentifierCoseKeyOpts;
function isManagedIdentifierOID4VCIssuerOpts(opts) {
    const { identifier } = opts;
    return ('method' in opts && opts.method === 'oid4vci-issuer') || (0, common_1.isOID4VCIssuerIdentifier)(identifier);
}
exports.isManagedIdentifierOID4VCIssuerOpts = isManagedIdentifierOID4VCIssuerOpts;
function isManagedIdentifierJwkOpts(opts) {
    const { identifier } = opts;
    return ('method' in opts && opts.method === 'jwk') || (0, common_1.isJwkIdentifier)(identifier);
}
exports.isManagedIdentifierJwkOpts = isManagedIdentifierJwkOpts;
function isManagedIdentifierX5cOpts(opts) {
    const { identifier } = opts;
    return ('method' in opts && opts.method === 'x5c') || (0, common_1.isX5cIdentifier)(identifier);
}
exports.isManagedIdentifierX5cOpts = isManagedIdentifierX5cOpts;
function isManagedIdentifierCoseKeyResult(object) {
    return object && typeof object === 'object' && 'method' in object && object.method === 'cose_key';
}
exports.isManagedIdentifierCoseKeyResult = isManagedIdentifierCoseKeyResult;
function isManagedIdentifierDidResult(object) {
    return object && typeof object === 'object' && 'method' in object && object.method === 'did';
}
exports.isManagedIdentifierDidResult = isManagedIdentifierDidResult;
function isManagedIdentifierX5cResult(object) {
    return object && typeof object === 'object' && 'method' in object && object.method === 'x5c';
}
exports.isManagedIdentifierX5cResult = isManagedIdentifierX5cResult;
function isManagedIdentifierJwkResult(object) {
    return object && typeof object === 'object' && 'method' in object && object.method === 'jwk';
}
exports.isManagedIdentifierJwkResult = isManagedIdentifierJwkResult;
function isManagedIdentifierKidResult(object) {
    return object && typeof object === 'object' && 'method' in object && object.method === 'kid';
}
exports.isManagedIdentifierKidResult = isManagedIdentifierKidResult;
function isManagedIdentifierKeyResult(object) {
    return object && typeof object === 'object' && 'method' in object && object.method === 'key';
}
exports.isManagedIdentifierKeyResult = isManagedIdentifierKeyResult;
