// @credo-ts/openid4vc/build/openid4vc-verifier/OpenId4VcVerifierModuleConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenId4VcVerifierModuleConfig = void 0;
const router_1 = require("../shared/router");
class OpenId4VcVerifierModuleConfig {
    constructor(options) {
        this.options = options;
        this.router = options.router ?? (0, router_1.importExpress)().Router();
    }
    get baseUrl() {
        return this.options.baseUrl;
    }
    get authorizationRequestEndpoint() {
        // Use user supplied options, or return defaults.
        const userOptions = this.options.endpoints?.authorizationRequest;
        return {
            ...userOptions,
            endpointPath: this.options.endpoints?.authorizationRequest?.endpointPath ?? '/authorization-requests',
        };
    }
    get authorizationEndpoint() {
        // Use user supplied options, or return defaults.
        const userOptions = this.options.endpoints?.authorization;
        return {
            ...userOptions,
            endpointPath: userOptions?.endpointPath ?? '/authorize',
        };
    }
}
exports.OpenId4VcVerifierModuleConfig = OpenId4VcVerifierModuleConfig;
