// @sphereon/ssi-sdk.core/dist/link-handlers/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LinkHandlerEventType = exports.emitLinkHandlerURLEvent = exports.DefaultLinkPriorities = void 0;
var DefaultLinkPriorities;
(function (DefaultLinkPriorities) {
    DefaultLinkPriorities[DefaultLinkPriorities["HIGHEST"] = 0] = "HIGHEST";
    DefaultLinkPriorities[DefaultLinkPriorities["DEFAULT"] = Number.MAX_VALUE / 2] = "DEFAULT";
    DefaultLinkPriorities[DefaultLinkPriorities["LOWEST"] = Number.MAX_VALUE] = "LOWEST";
})(DefaultLinkPriorities || (exports.DefaultLinkPriorities = DefaultLinkPriorities = {}));
const emitLinkHandlerURLEvent = (event, context) => {
    return context.agent.emit(LinkHandlerEventType.LINK_HANDLER_URL, event);
};
exports.emitLinkHandlerURLEvent = emitLinkHandlerURLEvent;
var LinkHandlerEventType;
(function (LinkHandlerEventType) {
    LinkHandlerEventType["LINK_HANDLER_URL"] = "link-handler-url";
})(LinkHandlerEventType || (exports.LinkHandlerEventType = LinkHandlerEventType = {}));
