// @sphereon/ssi-sdk.core/dist/loggers/eventLogger/EventLoggerBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const EventLogger_1 = __importDefault(require("./EventLogger"));
class EventLoggerBuilder {
    withContext(context) {
        this.context = context;
        return this;
    }
    withNamespace(namespace) {
        this.namespace = namespace;
        return this;
    }
    withSystem(system) {
        this.system = system;
        return this;
    }
    withSubSystem(subSystem) {
        this.subSystem = subSystem;
        return this;
    }
    withLogLevel(logLevel) {
        this.logLevel = logLevel;
        return this;
    }
    withInitiatorType(initiatorType) {
        this.initiatorType = initiatorType;
        return this;
    }
    build() {
        return new EventLogger_1.default({
            context: this.context,
            namespace: this.namespace,
            system: this.system,
            subSystem: this.subSystem,
            logLevel: this.logLevel,
            initiatorType: this.initiatorType,
        });
    }
}
exports.default = EventLoggerBuilder;
