// @sphereon/oid4vci-common/dist/types/OpenID4VCIVersions.types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultURISchemes = exports.OpenId4VCIVersion = void 0;
var OpenId4VCIVersion;
(function (OpenId4VCIVersion) {
    OpenId4VCIVersion[OpenId4VCIVersion["VER_1_0_08"] = 1008] = "VER_1_0_08";
    OpenId4VCIVersion[OpenId4VCIVersion["VER_1_0_09"] = 1009] = "VER_1_0_09";
    OpenId4VCIVersion[OpenId4VCIVersion["VER_1_0_11"] = 1011] = "VER_1_0_11";
    OpenId4VCIVersion[OpenId4VCIVersion["VER_1_0_12"] = 1012] = "VER_1_0_12";
    OpenId4VCIVersion[OpenId4VCIVersion["VER_1_0_13"] = 1013] = "VER_1_0_13";
    OpenId4VCIVersion[OpenId4VCIVersion["VER_UNKNOWN"] = Number.MAX_VALUE] = "VER_UNKNOWN";
})(OpenId4VCIVersion || (exports.OpenId4VCIVersion = OpenId4VCIVersion = {}));
var DefaultURISchemes;
(function (DefaultURISchemes) {
    DefaultURISchemes["INITIATE_ISSUANCE"] = "openid-initiate-issuance";
    DefaultURISchemes["CREDENTIAL_OFFER"] = "openid-credential-offer";
})(DefaultURISchemes || (exports.DefaultURISchemes = DefaultURISchemes = {}));
