// @sphereon/oid4vci-common/dist/types/Token.types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenError = exports.TokenErrorResponse = void 0;
var TokenErrorResponse;
(function (TokenErrorResponse) {
    TokenErrorResponse["invalid_request"] = "invalid_request";
    TokenErrorResponse["invalid_grant"] = "invalid_grant";
    TokenErrorResponse["invalid_client"] = "invalid_client";
    TokenErrorResponse["invalid_scope"] = "invalid_scope";
    TokenErrorResponse["invalid_dpop_proof"] = "invalid_dpop_proof";
})(TokenErrorResponse || (exports.TokenErrorResponse = TokenErrorResponse = {}));
class TokenError extends Error {
    constructor(statusCode, responseError, message) {
        super(message);
        this._statusCode = statusCode;
        this._responseError = responseError;
        // 👇️ because we are extending a built-in class
        Object.setPrototypeOf(this, TokenError.prototype);
    }
    get statusCode() {
        return this._statusCode;
    }
    get responseError() {
        return this._responseError;
    }
    getDescription() {
        return this.message;
    }
}
exports.TokenError = TokenError;
