// @credo-ts/openid4vc/build/openid4vc-verifier/repository/OpenId4VcVerifierRecord.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenId4VcVerifierRecord = void 0;
const core_1 = require("@credo-ts/core");
/**
 * For OID4VC you need to expos metadata files. Each issuer needs to host this metadata. This is not the case for DIDComm where we can just have one /didcomm endpoint.
 * So we create a record per openid issuer/verifier that you want, and each tenant can create multiple issuers/verifiers which have different endpoints
 * and metadata files
 * */
class OpenId4VcVerifierRecord extends core_1.BaseRecord {
    constructor(props) {
        super();
        this.type = OpenId4VcVerifierRecord.type;
        if (props) {
            this.id = props.id ?? core_1.utils.uuid();
            this.createdAt = props.createdAt ?? new Date();
            this._tags = props.tags ?? {};
            this.verifierId = props.verifierId;
        }
    }
    getTags() {
        return {
            ...this._tags,
            verifierId: this.verifierId,
        };
    }
}
exports.OpenId4VcVerifierRecord = OpenId4VcVerifierRecord;
OpenId4VcVerifierRecord.type = 'OpenId4VcVerifierRecord';
